src/nvim/buffer.c

```diff
@@ -3904,31 +3904,53 @@ int build_stl_str_hl(
         if (item[item_idx].type == Middle) {
           middle_mark = item_idx;
           break;
-          /*
-          // Move the statusline to make room for the middle characters
-          char_u *middle_end = item[item_idx].start + (maxwidth - width);
-          STRMOVE(middle_end, item[item_idx].start);
+        }
+      }
+
+      if (middle_mark > 0) {
+        // Handle a combination of middle mark and centering
+        //
+        // text\ %[this\ text\ here%=\ and\ some%]finishing
+        // text      his text here and some     finishing
+        // quad1             quad2|quad3            quad4
+        //       --fill1--                --fill2--
+        // int len_a = item[center_begin].start - out;
+        int len_b = item[middle_mark].start - out;
+        // int len_c = item[center_end].start - out;
+        // Not sure about this one
+        int len_d = width;
+
+        int fill_left = maxwidth / 2 - len_b;
+        int fill_right = maxwidth - maxwidth / 2 - len_d + len_b;
+
+        char_u *start = item[center_begin].start + fill_left;
+        STRMOVE(start, item[center_begin].start);
+
+        // Fillchars on the left side
+        for (char_u *s = item[center_begin].start; s < start; s++) {
+            *s = fillchar;
+        }
 
-          // Fill the middle section with our fill character
-          for (char_u *s = item[item_idx].start; s < middle_end; s++)
-            *s = '=';
+        int item_idx = center_begin;
+        // Adjust the offset of any items after the middle
+        for (item_idx++; item_idx <= center_end; item_idx++)
```

Comment:
This is a very uncommon style and might confuse reviewers when exactly `item_idx` is incremented. Please change it to:

```c
item_idx++;
for (; item_idx <= center_end; item_idx++)
```

...or simply with the line before...

```c
int item_idx = center_begin + 1;
for (; item_idx <= center_end; item_idx++)
```

Oh, and you forgot the parentheses for the `if`. ;-)